lib/adal-aad-node/src/engine/AzureAuth.ts

```diff
@@ -0,0 +1,469 @@
+import { ProviderSettings, SecureStorageProvider, Tenant, AADResource, LoginResponse, Deferred, AzureAccount, Logger, MessageDisplayer, ErrorLookup, CachingProvider, RefreshTokenPostData, AuthorizationCodePostData, TokenPostData, AccountKey, StringLookup, DeviceCodeStartPostData, DeviceCodeCheckPostData, AzureAuthType, AccountType } from '../models';
+import { AzureAuthError } from '../errors/azureAuthError';
+import { ErrorCodes }  from '../errors/errors';
+import axios, { AxiosRequestConfig, AxiosResponse } from 'axios';
+import { AccessToken, Token, TokenClaims, RefreshToken, OAuthTokenResponse } from '../models/auth';
+import * as qs from 'qs';
+import * as url from 'url';
+
+export abstract class AzureAuth {
+	public static readonly ACCOUNT_VERSION = '2.0';
+
+	protected readonly commonTenant: Tenant = {
+		id: 'common',
+		displayName: 'common'
+	};
+	protected readonly clientId: string;
+	protected readonly loginEndpointUrl: string;
+	constructor(
+		protected readonly providerSettings: ProviderSettings,
+		protected readonly secureStorage: SecureStorageProvider,
+		protected readonly cachingProvider: CachingProvider,
+		protected readonly logger: Logger,
+		protected readonly messageDisplayer: MessageDisplayer,
+		protected readonly errorLookup: ErrorLookup,
+		protected readonly userInteraction: UserInteraction,
+		protected readonly stringLookup: StringLookup,
+		protected readonly azureAuthType: AzureAuthType
+	) {
+		this.clientId = providerSettings.id;
+		this.loginEndpointUrl = providerSettings.loginEndpoint;
+	}
+
+	protected abstract async login(tenant: Tenant, resource: AADResource): Promise<LoginResponse>;
+
+	public async startLogin(): Promise<AzureAccount | undefined> {
+		let loginComplete: Deferred<void> | undefined;
+		try {
+			const result = await this.login(this.commonTenant, this.providerSettings.resources.windowsManagementResource);
+			loginComplete = result?.authComplete;
+			if (!result?.response) {
+				this.logger.error('Authentication failed');
+				return undefined;
+			}
+			const account = await this.hydrateAccount(result.response.accessToken, result.response.tokenClaims);
+			loginComplete?.resolve();
```

Comment:
There's a couple paths here that aren't resolve/rejecting the loginComplete promise - is that intentional?